### PR TITLE
Adding check for the triangle_3d_3 area calculation

### PR DESCRIPTION
**Pull request by loumalouomega**
I am having some troubles with interface triangles which are in a bad shape, and I saw that there is not any check in the Heron formula to check that. I switch to the math_utils Heron formula that includes a check

### CODE REVIEW

**Comment by loumalouomega** [comment]
Thing to discuss if the consider that this check is expensive or we can assume it (it is just a check if the area before sqrt is negative)

File: kratos/geometries/triangle_3d_3.h [hunk 1]
```diff
@@ -421,8 +424,9 @@ template<class TPointType> class Triangle3D3
      * :TODO: could be replaced by something more suitable
      * (comment by janosch)
      */
-    virtual double DomainSize() const override {
-      return Area();
+    virtual double DomainSize() const override 
+    {
+        return Area();
```
Comment (rubenzorrilla) [inline kratos/geometries/triangle_3d_3.h:429]: Maybe I'm missing something but why is the DomainSize() method returning the Area?

Comment (loumalouomega) [reply kratos/geometries/triangle_3d_3.h:429]: I don't know, it was that way when I arrived

Comment (pooyan-dadvand) [reply kratos/geometries/triangle_3d_3.h:429]: This is THE example of how a bad name for a method can mislead everyone. DomainSize returns area for 2d shapes and volume for 3d shapes

Comment (loumalouomega) [reply kratos/geometries/triangle_3d_3.h:429]: This is a 2D triangle in 3D space, what do we consider?

Comment (pooyan-dadvand) [reply kratos/geometries/triangle_3d_3.h:429]: is a 2D shape so has area

**Review by pooyan-dadvand — CHANGES_REQUESTED** [review]
Actually I don't agree with this change because in my mesh optimisation I may get a negative area within a step and is not a problem. So I would suggest you to add a debug check in your application

**Comment by loumalouomega** [comment]
I did the check in my application, I will change the flag

**Comment by loumalouomega** [comment]
@pooyan-dadvand Done

**Comment by RiccardoRossi** [comment]
i have no further comments on my side, let's wait for @pooyan-dadvand feedback

**Comment by rubenzorrilla** [comment]
@loumalouomega shall we close this?

**Comment by loumalouomega** [comment]
Yes, I will close this

**Comment by pooyan-dadvand** [comment]
Sorry for my absent.... Meanwhile the only important change in that branch was the check flag?

File: kratos/geometries/triangle_3d_3.h [hunk 2]
```diff
@@ -644,21 +658,22 @@ template<class TPointType> class Triangle3D3
...
+        const double sb = (b[0]*b[0])+(b[1]*b[1])+(b[2]*b[2]);
+        const double sc = (c[0]*c[0])+(c[1]*c[1])+(c[2]*c[2]);
 
-      // Shortest altitude is the one intersecting the largest base (or edge).
-      double base = CalculateMaxEdgeLength(sa,sb,sc);
+        // Shortest altitude is the one intersecting the largest base (or edge).
+        const double base = CalculateMaxEdgeLength(sa,sb,sc);
 
-      return normFactor * (Area() * 2 / base ) / base;
+        return normFactor * (Area() * 2 / base ) / base;
```
Comment (maceligueta) [inline kratos/geometries/triangle_3d_3.h:676]: One division can be avoided here. @loumalouomega

File: kratos/geometries/triangle_3d_3.h [hunk 3]
```diff
@@ -623,14 +636,15 @@ template<class TPointType> class Triangle3D3
...
 
-      double a = MathUtils<double>::Norm3(this->GetPoint(0)-this->GetPoint(1));
-      double b = MathUtils<double>::Norm3(this->GetPoint(1)-this->GetPoint(2));
-      double c = MathUtils<double>::Norm3(this->GetPoint(2)-this->GetPoint(0));
+        const double a = MathUtils<double>::Norm3(this->GetPoint(0)-this->GetPoint(1));
+        const double b = MathUtils<double>::Norm3(this->GetPoint(1)-this->GetPoint(2));
+        const double c = MathUtils<double>::Norm3(this->GetPoint(2)-this->GetPoint(0));
 
-      return normFactor * Area() / std::pow(a+b+c, 2);
+        return normFactor * Area() / std::pow(a+b+c, 2);
```
Comment (maceligueta) [inline kratos/geometries/triangle_3d_3.h:647]: std::pow(x, 2) might be replaced by x*x. Who knows if the optimizer is doing it already... @loumalouomega

**Comment by loumalouomega** [comment]
I thought that @pooyan-dadvand said that he doesn't need the check for some of his applications, and we agreed to put directly in my code a check to avoid the problem. That's why I decided to close the discussion.

**Comment by pooyan-dadvand** [comment]
You are right @loumalouomega. It gives problem to my remeshing.

**Comment by loumalouomega** [comment]
@pooyan-dadvand So, do we keep it closed?

**Comment by pooyan-dadvand** [comment]
Yes we keep this closed. @roigcarlo would you please take in account the comments of @maceligueta ?

**Comment by roigcarlo** [comment]
@pooyan-dadvand, @maceligueta With -O3 (Release) `sqrt::pow(x,2)` generates the same code as `x*x`. This pressumably also happens with -O2 (RelWithDebInfo). 

I havn't seen the implementation of `sqrt::pow` myself but I suspect they have some sort of specialization which would _probably_ make `sqrt::pow(x,2)` faster than a plain `x*x`. but anyway is going to be marginal.

So the only reason to rewrote that as `(a+b+c) * (a+b+c)` would be if we really need that the debug version of the code runs faster ( and I don't think that this operation is particulary expensive) or to improve readability.

Still, if it helps in any of those we can ofcourse change it.

**Edit:**

Also notice that it gets harder to read due to the extra brackets

```C++
return normFactor * Area() / std::pow(a+b+c, 2);
return normFactor * Area() / ((a+b+c) * (a+b+c));
```

**Comment by pooyan-dadvand** [comment]
If you are sure about the `pow `I am ok with that. Nevertheless you can use an auxiliary variable to improve readability:
```c++
const double sum = a + b + c;
return  normFactor * Area() / (sum * sum);
```
For me the more critical one is the double divisions:
```c++
return normFactor * (Area() * 2 / base ) / base;
```

**Pull request closed without merging**